JDK8 Win64 java -version crash VMState: 0x000503ff Shared cache offset out of bounds

Happens only on JDK8 64bit (internal nightly build). Seems to be 100% reproducible. I have a set of diagnostic files saved. Last passed build was Nov 10 58d490796d0276f5c3f3b45a6cf21513ecf7bde1 
```
00:01:43.811  unzip file: OpenJ9-JDK8-x86-64_windows-20191121-211505.tar.gz ...
00:01:58.863  unzip file: test-images.tar.gz ...
00:02:01.630  Run C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly/openjdkbinary/j2sdk-image/bin/java -version
00:02:01.630  Assertion failed at C:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Nightly\build\windows-x86_64-normal-server-release\vm\compiler\env\J9SharedCache.cpp:457: false
00:02:01.630  VMState: 0x000503ff
00:02:01.630  	Shared cache offset out of bounds
00:02:01.630  compiling java/util/Hashtable.put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object; at level: cold
00:02:01.630  
00:02:01.630  Unhandled exception
00:02:01.630  Type=Segmentation error vmState=0x000503ff
00:02:01.630  Windows_ExceptionCode=c0000005 J9Generic_Signal=00000004 ExceptionAddress=0000000057F18A46 ContextFlags=0010001f
00:02:01.630  Handler1=0000000057C1B470 Handler2=00007FF87553CC70 InaccessibleWriteAddress=0000000000000000
00:02:01.630  RDI=000000000E868788 RSI=000000000E868780 RAX=0000000000000000 RBX=000A42E7ED469368
00:02:01.630  RCX=A1AEBD1D8CAD0000 RDX=0000000000000000 R8=0000000000000069 R9=00000000000000CB
00:02:01.630  R10=00000000000002C0 R11=0000000000000059 R12=0000000000000003 R13=FFFFFFFFF1797878
00:02:01.630  R14=000000000E6E7DB0 R15=0000000001824300
00:02:01.630  RIP=0000000057F18A46 RSP=000000000E8685E0 RBP=000000000E8687C8 GS=002B
00:02:01.630  FS=0053 ES=002B DS=002B
00:02:01.630  XMM0 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM1 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM2 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM3 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM4 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM5 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM6 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM7 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM8 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM9 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM10 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM11 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM12 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM13 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM14 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  XMM15 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:02:01.631  Module=C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly\openjdkbinary\j2sdk-image\jre\bin\compressedrefs\j9jit29.dll
00:02:01.631  Module_base_address=0000000057BD0000 Offset_in_DLL=0000000000348a46
00:02:01.631  
00:02:01.631  Method_being_compiled=java/util/Hashtable.put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
00:02:01.631  Target=2_90_20191121_232 (Windows Server 2016 10.0 build 14393)
00:02:01.631  CPU=amd64 (4 logical CPUs) (0x1fff8d000 RAM)
00:02:01.631  ----------- Stack Backtrace -----------
00:02:01.631  openjdk version "1.8.0_232-internal"
00:02:01.631  OpenJDK Runtime Environment (build 1.8.0_232-internal-jenkins_2019_11_21_18_40-b00)
00:02:01.631  Eclipse OpenJ9 VM (build ibm_sdk-b2a72599ec, JRE 1.8.0 Windows Server 2016 amd64-64-Bit Compressed References 20191121_232 (JIT enabled, AOT enabled)
00:02:01.631  OpenJ9   - b2a72599ec
00:02:01.631  OMR      - 8eb476a6e
00:02:01.631  JCL      - f6352f08a86 based on jdk8u232-b09)
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x2b65f6 (0x0000000057F18A46 [j9jit29+0x348a46])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x2b686d (0x0000000057F18CBD [j9jit29+0x348cbd])
00:02:02.506  j9jit_testarossa+0x7e84 (0x0000000057C47C94 [j9jit29+0x77c94])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x9a698 (0x0000000057CFCAE8 [j9jit29+0x12cae8])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x9e338 (0x0000000057D00788 [j9jit29+0x130788])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x9834d (0x0000000057CFA79D [j9jit29+0x12a79d])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x9caa2 (0x0000000057CFEEF2 [j9jit29+0x12eef2])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x9b4b3 (0x0000000057CFD903 [j9jit29+0x12d903])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0xa3ff2 (0x0000000057D06442 [j9jit29+0x136442])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x804c6 (0x0000000057CE2916 [j9jit29+0x112916])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x35cd6e (0x0000000057FBF1BE [j9jit29+0x3ef1be])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x35f0aa (0x0000000057FC14FA [j9jit29+0x3f14fa])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x3559d9 (0x0000000057FB7E29 [j9jit29+0x3e7e29])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x352cb1 (0x0000000057FB5101 [j9jit29+0x3e5101])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x3553a2 (0x0000000057FB77F2 [j9jit29+0x3e77f2])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x3f9c39 (0x000000005805C089 [j9jit29+0x48c089])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x3fa766 (0x000000005805CBB6 [j9jit29+0x48cbb6])
00:02:02.506  Java_java_lang_invoke_MutableCallSite_invalidate+0x259dff (0x0000000057EBC24F [j9jit29+0x2ec24f])
00:02:02.506  (0x0000000057C26786 [j9jit29+0x56786])
00:02:02.506  (0x0000000057C2828E [j9jit29+0x5828e])
00:02:02.506  j9port_init_library+0x1b476 (0x00007FF87553DE46 [j9prt29+0x1de46])
00:02:02.506  j9port_init_library+0x1b674 (0x00007FF87553E044 [j9prt29+0x1e044])
00:02:02.506  (0x0000000057C28B2C [j9jit29+0x58b2c])
00:02:02.506  (0x0000000057C290F2 [j9jit29+0x590f2])
00:02:02.506  (0x0000000057C23EDA [j9jit29+0x53eda])
00:02:02.506  (0x0000000057C26E8E [j9jit29+0x56e8e])
00:02:02.506  j9port_init_library+0x1b6af (0x00007FF87553E07F [j9prt29+0x1e07f])
00:02:02.506  (0x0000000057C286DD [j9jit29+0x586dd])
00:02:02.506  omrthread_detach+0x1fd (0x00007FF878E85D6D [J9THR29+0x5d6d])
00:02:02.506  _endthreadex+0x43 (0x00000000599E1D9F [msvcr100+0x21d9f])
00:02:02.506  _endthreadex+0xdf (0x00000000599E1E3B [msvcr100+0x21e3b])
00:02:02.506  BaseThreadInitThunk+0x14 (0x00007FF87FB184D4 [KERNEL32+0x84d4])
00:02:02.506  RtlUserThreadStart+0x21 (0x00007FF88023E851 [ntdll+0x6e851])
00:02:02.506  ---------------------------------------
00:02:02.506  JVMDUMP039I Processing dump event "gpf", detail "" at 2019/11/22 01:34:14 - please wait.
00:02:02.506  JVMDUMP032I JVM requested System dump using 'C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly\openjdkbinary\core.20191122.013414.15288.0001.dmp' in response to an event
00:02:10.626  JVMDUMP010I System dump written to C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly\openjdkbinary\core.20191122.013414.15288.0001.dmp
00:02:10.626  JVMDUMP032I JVM requested Java dump using 'C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly\openjdkbinary\javacore.20191122.013414.15288.0002.txt' in response to an event
00:02:10.626  JVMDUMP010I Java dump written to C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly\openjdkbinary\javacore.20191122.013414.15288.0002.txt
00:02:10.626  JVMDUMP032I JVM requested Snap dump using 'C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly\openjdkbinary\Snap.20191122.013414.15288.0003.trc' in response to an event
00:02:10.626  JVMDUMP010I Snap dump written to C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly\openjdkbinary\Snap.20191122.013414.15288.0003.trc
00:02:10.626  JVMDUMP007I JVM Requesting JIT dump using 'C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Nightly\openjdkbinary\jitdump.20191122.013414.15288.0004.dmp'
```

Still seeing this crash on 3/4 test targets
based on 5ea61ceb3d14993e11703dd2b6d4080a2290f087
/job/Pipeline-Build-Test-Personal/4562/
```
00:06:07.949  compile:
00:06:07.949       [echo] Ant version is Apache Ant(TM) version 1.10.5 compiled on July 10 2018
00:06:07.949       [echo] ============COMPILER SETTINGS============
00:06:07.949       [echo] ===fork:                         yes
00:06:07.949       [echo] ===debug:                        on
00:06:08.824      [javac] Compiling 13 source files to C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\bin
00:06:09.683      [javac] Unhandled exception
00:06:09.683      [javac] Type=Segmentation error vmState=0x000503ff
00:06:09.683      [javac] Windows_ExceptionCode=c0000005 J9Generic_Signal=00000004 ExceptionAddress=0000000061A09C49 ContextFlags=0010001f
00:06:09.683      [javac] Handler1=00000000619CBBA0 Handler2=00007FFFB6ACCCF0 InaccessibleReadAddress=0000000082A2F189
00:06:09.683      [javac] RDI=000000000229A300 RSI=000000000F98E830 RAX=000000000F98E888 RBX=00007FF695800000
00:06:09.683      [javac] RCX=0000000000000000 RDX=0000000001BC46A8 R8=0000000000000001 R9=0000000000000004
00:06:09.683      [javac] R10=00007FFFB6C4CB28 R11=000000000F988960 R12=000000000F265BB0 R13=0000000063416261
00:06:09.683      [javac] R14=0000000082A2F18B R15=00000000000000A0
00:06:09.683      [javac] RIP=0000000061A09C49 RSP=000000000F988970 RBP=000000001F618F20 GS=002B
00:06:09.683      [javac] FS=0053 ES=002B DS=002B
00:06:09.683      [javac] XMM0 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.683      [javac] XMM1 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.683      [javac] XMM2 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.683      [javac] XMM3 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.683      [javac] XMM4 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.683      [javac] XMM5 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.683      [javac] XMM6 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM7 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM8 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM9 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM10 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM11 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM12 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM13 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM14 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] XMM15 0000000000000000 (f: 0.000000, d: 0.000000e+000)
00:06:09.684      [javac] Module=C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdkbinary\j2sdk-image\jre\bin\compressedrefs\j9jit29.dll
00:06:09.684      [javac] Module_base_address=0000000061980000 Offset_in_DLL=0000000000089c49
00:06:09.684      [javac] 
00:06:09.684      [javac] Method_being_compiled=sun/net/www/protocol/jar/Handler.parseContextSpec(Ljava/net/URL;Ljava/lang/String;)Ljava/lang/String;
00:06:09.684      [javac] Target=2_90_20200113_281 (Windows Server 2016 10.0 build 14393)
00:06:09.684      [javac] CPU=amd64 (4 logical CPUs) (0x1fff8d000 RAM)
00:06:09.684      [javac] ----------- Stack Backtrace -----------
00:06:09.684      [javac] Java_java_lang_invoke_InterfaceHandle_convertITableIndexToVTableIndex+0x439 (0x0000000061A09C49 [j9jit29+0x89c49])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x9d9e2 (0x0000000061AB0522 [j9jit29+0x130522])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0xa1658 (0x0000000061AB4198 [j9jit29+0x134198])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x9b99a (0x0000000061AAE4DA [j9jit29+0x12e4da])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x36304a (0x0000000061D75B8A [j9jit29+0x3f5b8a])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x3650ea (0x0000000061D77C2A [j9jit29+0x3f7c2a])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x35ba09 (0x0000000061D6E549 [j9jit29+0x3ee549])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x358ce1 (0x0000000061D6B821 [j9jit29+0x3eb821])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x35b3d2 (0x0000000061D6DF12 [j9jit29+0x3edf12])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x3feff9 (0x0000000061E11B39 [j9jit29+0x491b39])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x3ffb26 (0x0000000061E12666 [j9jit29+0x492666])
00:06:09.684      [javac] Java_java_lang_invoke_MutableCallSite_invalidate+0x26094b (0x0000000061C7348B [j9jit29+0x2f348b])
00:06:09.684      [javac] (0x00000000619D6E96 [j9jit29+0x56e96])
00:06:09.684      [javac] (0x00000000619D8A1B [j9jit29+0x58a1b])
00:06:09.684      [javac] j9port_init_library+0x1b4f6 (0x00007FFFB6ACDEC6 [j9prt29+0x1dec6])
00:06:09.684      [javac] j9port_init_library+0x1b6f4 (0x00007FFFB6ACE0C4 [j9prt29+0x1e0c4])
00:06:09.684      [javac] (0x00000000619D92BC [j9jit29+0x592bc])
00:06:09.684      [javac] (0x00000000619D9882 [j9jit29+0x59882])
00:06:09.684      [javac] (0x00000000619D45EA [j9jit29+0x545ea])
00:06:09.684      [javac] (0x00000000619D759E [j9jit29+0x5759e])
00:06:09.684      [javac] j9port_init_library+0x1b72f (0x00007FFFB6ACE0FF [j9prt29+0x1e0ff])
00:06:09.684      [javac] (0x00000000619D8E6D [j9jit29+0x58e6d])
00:06:09.684      [javac] omrthread_detach+0x1fd (0x00007FFFBF375D6D [J9THR29+0x5d6d])
00:06:09.684      [javac] _endthreadex+0x43 (0x00000000624E1D9F [MSVCR100+0x21d9f])
00:06:09.684      [javac] _endthreadex+0xdf (0x00000000624E1E3B [MSVCR100+0x21e3b])
00:06:09.684      [javac] BaseThreadInitThunk+0x14 (0x00007FFFCE5284D4 [KERNEL32+0x84d4])
00:06:09.684      [javac] RtlUserThreadStart+0x21 (0x00007FFFCF5BE851 [ntdll+0x6e851])
00:06:09.684      [javac] ---------------------------------------
00:06:09.684      [javac] JVMDUMP039I Processing dump event "gpf", detail "" at 2020/01/13 18:13:23 - please wait.
00:06:09.684      [javac] JVMDUMP032I JVM requested System dump using 'C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\scripts\core.20200113.181323.12108.0001.dmp' in response to an event
00:06:13.528      [javac] JVMDUMP010I System dump written to C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\scripts\core.20200113.181323.12108.0001.dmp
00:06:13.528      [javac] JVMDUMP032I JVM requested Java dump using 'C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\scripts\javacore.20200113.181323.12108.0002.txt' in response to an event
00:06:13.529      [javac] JVMDUMP010I Java dump written to C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\scripts\javacore.20200113.181323.12108.0002.txt
00:06:13.529      [javac] JVMDUMP032I JVM requested Snap dump using 'C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\scripts\Snap.20200113.181323.12108.0003.trc' in response to an event
00:06:13.529      [javac] JVMDUMP010I Snap dump written to C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\scripts\Snap.20200113.181323.12108.0003.trc
00:06:13.529      [javac] JVMDUMP007I JVM Requesting JIT dump using 'C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\scripts\jitdump.20200113.181323.12108.0004.dmp'
00:06:13.529      [javac] JVMDUMP010I JIT dump written to C:\Users\jenkins\workspace\Test_openjdk8_j9_extended.functional_x86-64_windows_Personal\openjdk-tests\TKG\scripts\jitdump.20200113.181323.12108.0004.dmp
00:06:13.529      [javac] JVMDUMP013I Processed dump event "gpf", detail "".
00:06:13.529  
00:06:13.529  BUILD FAILED
```

--- a/runtime/compiler/control/CompilationThread.cpp
+++ b/runtime/compiler/control/CompilationThread.cpp
@@ -5931,18 +5931,14 @@ void *TR::CompilationInfo::compileOnSeparateThread(J9VMThread * vmThread, TR::Il
          //
          if (vmThread->javaVM->sharedClassConfig->existsCachedCodeForROMMethod(vmThread, J9_ROM_METHOD_FROM_RAM_METHOD(method)))
             {
-            TR_J9SharedCacheVM *fe = (TR_J9SharedCacheVM *) TR_J9VMBase::get(jitConfig, vmThread, TR_J9VMBase::AOT_VM);
-            if (
-               static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader == TR_yes
-               || (
-                  static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader == TR_maybe
-                  && _sharedCacheReloRuntime.validateAOTHeader(fe, vmThread)
-                  )
-               )
+            if (static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader == TR_yes)
                {
                methodIsInSharedCache = TR_yes;
-//             if (getPersistentInfo()->isClassLoadingPhase() || TR::Options::getCmdLineOptions()->getOption(TR_ForceLoadAOT))
-                  useCodeFromSharedCache = true;
+               useCodeFromSharedCache = true;
+               }
+            else
+               {
+               TR_ASSERT_FATAL(static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader != TR_maybe, "Should not be possible for aotValidHeader to be TR_maybe at this point\n");
                }
             }
          }
@@ -11757,10 +11753,7 @@ TR::CompilationInfo::storeAOTInSharedCache(
       }
    else if (static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader == TR_maybe)
       {
-      // If validation has been performed, then a header already existed
-      // or one was already been created in this JVM
-      TR_J9SharedCacheVM *fe = (TR_J9SharedCacheVM *) TR_J9VMBase::get(jitConfig, vmThread, TR_J9VMBase::AOT_VM);
-      safeToStore = entry->_compInfoPT->reloRuntime()->storeAOTHeader(fe, vmThread);
+      TR_ASSERT_FATAL(static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader != TR_maybe, "Should not be possible for aotValidHeader to be TR_maybe at this point\n");
       }
    else
       {

--- a/runtime/compiler/control/rossa.cpp
+++ b/runtime/compiler/control/rossa.cpp
@@ -1868,13 +1868,40 @@ aboutToBootstrap(J9JavaVM * javaVM, J9JITConfig * jitConfig)
 #if defined(J9VM_OPT_SHARED_CLASSES)
    if (isSharedAOT)
       {
-      /* If AOT Shared Classes is turned ON, perform compatibility checks for AOT Shared Classes */
-      if (0) //!validateSharedClassAOTHeader(javaVM, curThread, compInfo))
+      bool validateSCC = true;
+
+#if defined(J9VM_OPT_JITSERVER)
+      if (compInfo->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
+         validateSCC = false;
+#endif
+
+      if (validateSCC)
          {
-         TR::Options::getAOTCmdLineOptions()->setOption(TR_NoLoadAOT);
-         TR::Options::getAOTCmdLineOptions()->setOption(TR_NoStoreAOT);
-         TR::Options::setSharedClassCache(false);
-         TR_J9SharedCache::setSharedCacheDisabledReason(TR_J9SharedCache::AOT_DISABLED);
+         /* If AOT Shared Classes is turned ON, perform compatibility checks for AOT Shared Classes
+          *
+          * This check has to be done after latePostProcessJIT so that all the necessary JIT options
+          * can be set
+          */
+         TR_J9VMBase *fe = TR_J9VMBase::get(jitConfig, curThread);
+         if (!compInfo->reloRuntime()->validateAOTHeader(fe, curThread))
+            {
+            TR_ASSERT_FATAL(static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader != TR_yes,
+                            "aotValidHeader is TR_yes after failing to validate AOT header\n");
+
+            /* If this is the second run, then failing to validate AOT header will cause aotValidHeader
+             * to be TR_no, in which case the SCC is not valid for use. However, if this is the first
+             * run, then aotValidHeader will be TR_maybe; try to store the AOT Header in this case.
+             */
+            if (static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader == TR_no
+                || !compInfo->reloRuntime()->storeAOTHeader(fe, curThread))
+               {
+               static_cast<TR_JitPrivateConfig *>(jitConfig->privateConfig)->aotValidHeader = TR_no;
+               TR::Options::getAOTCmdLineOptions()->setOption(TR_NoLoadAOT);
+               TR::Options::getAOTCmdLineOptions()->setOption(TR_NoStoreAOT);
+               TR::Options::setSharedClassCache(false);
+               TR_J9SharedCache::setSharedCacheDisabledReason(TR_J9SharedCache::AOT_DISABLED);
+               }
+            }
          }
 
       if (TR::Options::getAOTCmdLineOptions()->getOption(TR_NoStoreAOT))

